[lit][ci] Enable testing Release Lit workflow in forks

nightlark opened the pull request:
I'm looking at Python packaging of some parts of LLVM and updating them to use current best practices (https://github.com/llvm/llvm-project/issues/125220#issuecomment-2635940515), and couldn't test the CI release workflow in my fork.

This updates the CI workflow to:
* Enable building in forks without erroring on the permissions/upload steps by only running those steps when the workflow is running in the repo within the main LLVM org
* Add an option to save the built package as an artifact to download for local testing/inspection
* Add a dry-run option for testing the release workflow within the main LLVM repository without actually doing the upload (also as an input for workflow_call, so in the future the release workflow that releases everything can also be set up to do a test of the release workflows without uploading anything.)

These things could be split out into separate PRs if desired. The first two are probably the key ones needed to facilitate testing in forks.

nightlark committed:
[lit][ci] Enable testing Release Lit workflow in forks

Signed-off-by: Ryan Mast <mast.ryan@gmail.com>

boomanaiden154 reviewed (commented):
A couple notes:
1. We should ideally make this workflow run in dry run mode when a pull request is made that touches this workflow file, and maybe on push. Then we can drop the explicit dry-run input. See `build-ci-container.yml` for an example.
2. We should save the artifacts unconditionally.
3. I'm not a fan of adding `github.repository_owner == 'llvm'` lines per-step. According to https://llvm.org/docs/CIBestPractices.html, we should have a job level `github.repository_owner = 'llvm'` tag. I think enabling testing on PRs makes this less of a burden on testing.

nightlark commented:
Since a squashing workflow is used, should I also edit the PR description to have less commentary and just be what the final commit message body should say in it?

nightlark commented:
> A couple notes:
> 
> 1. We should ideally make this workflow run in dry run mode when a pull request is made that touches this workflow file, and maybe on push. Then we can drop the explicit dry-run input. See `build-ci-container.yml` for an example.

I removed the explicit dry-run input, and used a check against `contains(github.workflow_ref, '/.github/workflows/release-tasks.yml')` to limit the permission check and upload steps to the release-tasks workflow triggering it (otherwise the permission check fails on PRs, and the package doesn't get built or tested). It seems currently for actual releases, this workflow will only get triggered by release-tasks.yml when a `llvmorg-` tag is pushed, and I'm guessing we don't want to add any new conditions that can trigger an actual release.

I also added the `lit` package folder to the set of paths that can trigger running the workflow, to validate that the check-list tests still pass and any packaging related changes are okay.

Somewhat annoying, for workflow_call triggers almost all of the information in the github context is inherited from the parent workflow -- in this case that means the event_name is `push`, which conflicts with doing a "dry-run" test for commits to main. `github.workflow_ref` is the only real indication I found to tell if this workflow is being triggered by a workflow_call and not a push to main, hence using `contains` for the check. An alternative is modifying release-tasks.yml to pass in an input to use as a flag.

> 2. We should save the artifacts unconditionally.

Done.

> 3. I'm not a fan of adding `github.repository_owner == 'llvm'` lines per-step. According to https://llvm.org/docs/CIBestPractices.html, we should have a job level `github.repository_owner = 'llvm'` tag. I think enabling testing on PRs makes this less of a burden on testing.

I moved the repository owner check to the top of the workflow, along with allowing `workflow_dispatch` as a means to still manually trigger a dry run in forks (which is needed since I don't have the permissions necessary to let the workflows run on my PR in the LLVM repository so otherwise testing subsequent changes requires constant nagging of someone to approve the workflow run on the PR).
